### PR TITLE
821: Manual create

### LINKED ISSUE #821: Create entry view and routing
## Requirements
* Tapping the plus button from the entry list (android) opens the new entry screen
* Once saved, the user is routed to the read-only detail view (no toast)
* Make sure the item list is synced so that the new entry will be seen there too
* Tapping the 'X' button will cancel the creation of a new credential #823 
* Field validation and errors served on required fields #822 

## Visual Design
Invision: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/screens/344499818

Zeplin: https://app.zeplin.io/project/5b6895bfe4af825140aa8dbc/screen/5c548fc35f8c1d3577504065

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #821 

## Testing and Review Notes
Designs: https://app.zeplin.io/project/5b6895bfe4af825140aa8dbc?seid=5c5490093af38330875be09c 

Routes back to the item list after a successful save. Might do the routing to the newly created item on #822.

## Screenshots or Videos
create_new_entry

image image

### eng todo

- [x] `+` button
- [x] new login fragment
- [x] routes to/from itemlist, locked screen, item detail
- [x] feature flag
- [x] unit tests
- [x] hook up toggle password button
- [x] ripple on `+` button click

### other

- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [ ] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [ ] add unit tests
- [ ] request the "UX" team perform a design review (if/when applicable)
- [ ] make sure CI builds are passing (e.g.: fix lint and other errors)
- [ ] check on the [accessibility](https://mozilla-lockwise.github.io/lockwise-android/accessibility/) of any added UI

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/presenter/AppRoutePresenter.kt [hunk 1]
```diff
@@ -142,13 +143,15 @@ class AppRoutePresenter(
 
             R.id.fragment_locked to R.id.fragment_item_list -> R.id.action_locked_to_itemList
             R.id.fragment_locked to R.id.fragment_welcome -> R.id.action_locked_to_welcome
+            R.id.fragment_locked to R.id.fragment_create -> R.id.action_locked_to_manualCreate
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/AppRoutePresenter.kt]: If the app autolocks in the middle of create, do we want to discard everything? My assumption is that we would want to continue where we left off and continue from the create screen until we intentionally :x: or ✅

**Comment by eliserichards** [comment]
I found the `Item` in `EditItemFragment` to be redundant. My preference would be to remove `Item` from the names altogether and stick with: 

`Create` + View, Fragment, Presenter.
`Edit` + View, Fragment, Presenter.
`ItemList` + View, Fragment, Presenter
`ItemDetail` + View, Fragment, Presenter

**Comment by eliserichards** [comment]
We arrived on the following for naming conventions:

`CreateItem` + View, Fragment, Presenter.
`EditItem` + View, Fragment, Presenter.
`DisplayItem` + View, Fragment, Presenter
`ItemList` + View, Fragment, Presenter

**Comment by eliserichards** [comment]
@jhugman this needs some styling work. Looks like the `view` isn't applying the correct styles or managing the `closeKeyboard` functions the way that it should after I refactored the presenters. I would like a pair of eyes on this, or maybe an hour or so to ✨ pair ✨ on Monday if you have some bandwidth 😄

**Comment by jhugman** [comment]
> I would like a pair of eyes on this, or maybe an hour or so to ✨ pair ✨ on Monday if you have some bandwidth 😄

We can do both. :)

File: app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt [hunk 2]
```diff
@@ -131,6 +132,11 @@ class ItemListPresenter(
             .subscribe(dispatcher::dispatch)
             .addTo(compositeDisposable)
 
+        view.createNewEntryClick
+            .map { RouteAction.Create }
```
Comment (jhugman) [inline app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt]: This should match the new naming scheme.

**Comment by eliserichards** [comment]
- [x] Rename route actions to match new naming scheme for display/edit/create
- [ ] order of calling `super.x` on new presenters
- [ ] styles not set correctly
- [ ] close keyboard from `onPause` `onDestroyView` not working anymore

**Comment by eliserichards** [comment]
@jhugman I fixed the FAB issue, but I still have two outstanding issues (see my checklist in the last comment). Other than that the functionalities seem to be working as expected.

**Comment by changecourse** [comment]
Moving a conversation I had with @jhugman in slack to this PR: 

essentially, in almost all use cases, a user shouldn't see these inline errors until they've actually made a misstep. I think the error messages should only be visible once someone leaves focus, and we've checked that they hadn't met the requirements. 

So as a use case, someone comes to this screen after hitting the + button...

1. no errors yet present, field focused on web address, keyboard expanded
2. upon entering a web address and navigating to the next field (username) we do a client-side check on the web address. if it is not meeting requirements we then throw the client-side error to the web address field
3. same sort of logic for other fields... we assume that someone is not going to type a duplicate or leave a field blank, so we don't show any errors until they've actively demonstrated that they have ignored requirements, by navigating into AND out of a given field

**Comment by jhugman** [comment]
@changecourse I'd like to move this into a follow up issue. This PR is big enough and long running enough to get this landed before more work goes into it.

**Pull request merged**